text/0323-array-functions.md

```diff
@@ -0,0 +1,130 @@
+- Start Date: 2018-03-25
+- RFC PR: https://github.com/emberjs/rfcs/pull/323
+- Ember Issue: (leave this empty)
+
+# Array functions
+
+## Summary
+
+This RFC proposes a way to work with arrays in ember without requiring that we extend the global `Array` prototype or manually wrap arrays with `Ember.A(array)`.
+
+Specifically, for each method on the `MutableArray` mixin we will provide a similarly named function that takes an array as its first argument followed by the arguments to the method.
+
+For example, instead of calling `array.pushObject(item)` you can call `pushObject(array, item)`. For a full list see the expandible list in [Detailed design](#detailed-design).
+
+## Motivation
+
+Currently it is not a great experience to write an app with `Array` prototype extensions disabled. You end up needing to wrap arrays everywhere with `Ember.A(array)` which loops through each method on the `NativeArray` mixin and assigns it as an enumerable own property on the array.
+
+The are several problems here:
+
+- It's annoying to defensively write `Ember.A()` around every array
+- It's wasteful to looping through all the array methods and assign them all. Especially if you're only going to use one of them
+- Iterating over an array with `for..in` or `Object.keys` will not behave as generally expected because the methods are assigned enumerably (for performance reasons)
+
+All of these problems go away when using array functions instead of array methods.
+
+It's also worth noting that when you using fastboot you are required to have `Array` prototype extensions disabled currently and thus forced to deal with these issues.
+
+## Detailed design
+
+For each method in the `MutableArray` mixin we will export a similarly named function from `@ember/array`.
+
+For example, the method `array.objectAt(index)` will map to the function `objectAt(array, index)` with implementation
+
+```js
+function objectAt(array, index) {
+  if (Array.isArray(array)) {
+    return array[index];
+  } else {
+    return array.objectAt(index);
+  }
+}
+```
+
+As you can see, the native array case is handled explicitly in order to avoid assuming that the global `Array` prototype has been extended, but otherwise we just delegate to the `objectAt` method.
+
+For reference, below is an expandable list of all array functions we will need.
+
+<details>
+  <summary>List of array functions (49)</summary>
+  <code>
+    <ul>
+      <li>addArrayObserver</li>
+      <li>addObject</li>
+      <li>addObjects</li>
+      <li>any</li>
+      <li>arrayContentDidChange</li>
+      <li>arrayContentWillChange</li>
+      <li>clear</li>
+      <li>compact</li>
+      <li>every</li>
```

Comment:
Does it make sense to port all available methods? Methods like `every`, `filter` or `find` are available in all supported browsers now, so can't we just get rid of them?

Reply:
I should list this as (yet another) alternative. Initially I thought the same as you, but then I decided to go with the current proposal to keep things "simpler", e.g. to avoid gotchas like (why can I import `mapBy` but not `map`? Both are documented as methods on the Array mixin).

It's also possible to go to the opposite end of the spectrum and only include `objectAt` and `replace`, but see https://github.com/emberjs/rfcs/pull/323#discussion_r176943275

Reply:
If you think about building a set of changes with the proposed `|>` operator, having all of the functions (like `filter` and `find`) available with the same semantics allows for better consistency

```
users
|> filter(u => isAdmin(u))
|> mapBy('name')
```